src/io/binary_parser.cpp

```diff
@@ -156,32 +156,84 @@ long long read_variables(string filename, dd::FactorGraph &fg)
     return count;
 }
 
-long long read_factors(string filename, dd::FactorGraph &fg)
+long long read_factors(string filename, dd::FactorGraph &fg, int inc)
 {
-    ifstream file;
-    file.open(filename.c_str(), ios::in | ios::binary);
-    long long count = 0;
-    long long id;
-    long long weightid;
-    short type;
-    long long edge_count;
-    while (file.good()) {
-        file.read((char *)&id, 8);
-        file.read((char *)&weightid, 8);
-        file.read((char *)&type, 2);
-        if (!file.read((char *)&edge_count, 8)) break;
+    if (inc) {
```

Comment:
This big `if (inc)` is confusing. It's not immediately obvious what the difference is inside `if (inc)` and `else`. The two parts of code look similar. Is there a clearer way to rewrite these (less duplication)? If not, I think it's better to just have two functions, one for inc and the other for non-inc. Also, a comment would be very helpful. It's hard to figure out what's going on when later looking at the code...